b2sdk/sync/folder.py

```diff
@@ -10,17 +10,26 @@
 
 import logging
 import os
+import platform
+import re
 import six
 import sys
 
 from abc import ABCMeta, abstractmethod
 from b2sdk.exception import CommandError
-from .exception import EnvironmentEncodingError
+from .exception import EnvironmentEncodingError, UnSyncableFilename
 from .file import File, FileVersion
 from .scan_policies import DEFAULT_SCAN_MANAGER
 from ..raw_api import SRC_LAST_MODIFIED_MILLIS
 from ..utils import fix_windows_path_limit, is_file_readable
 
+DRIVE_MATCHER = re.compile(r"^([A-Za-z]):([/\\])")
+ABSOLUTE_PATH_MATCHER = re.compile(r"^(/)|^(\\)")
+RELATIVE_PATH_MATCHER = re.compile(
+    r"^(\.\.[/\\])|^(\.[/\\])|([/\\]\.\.[/\\])|([/\\]\.[/\\])|([/\\]\.\.)$|([/\\]\.)$|^(\.\.)$|" +
+    r"([/\\][/\\])|^(\.)$"
+)
```

Comment:
```suggestion
RELATIVE_PATH_MATCHER = re.compile(
                           # "abc" and "xyz" represent anything, including "nothing"
    r"^(\.\.[/\\])|" +     # ../abc or ..\abc
    r"^(\.[/\\])|" +       # ./abc or .\abc
    r"([/\\]\.\.[/\\])|" + # abc/../xyz or abc\..\xyz or abc\../xyz or abc/..\xyz
    r"([/\\]\.[/\\])|" +   # abc/./xyz or abc\.\xyz or abc\./xyz or abc/.\xyz
    r"([/\\]\.\.)$|" +     # abc/.. or abc\..
    r"([/\\]\.)$|" +       # abc/. or abc\. 
    r"^(\.\.)$|" +         # just ".."
    r"([/\\][/\\])|" +     # abc\/xyz or abc/\xyz or abc//xyz or abc\\xyz
    r"^(\.)$"              # just "."
)  # yapf: disable
```
I have translated the regex into somewhat more readable patterns and now that I look at it, it seems that it will not work correctly in some cases. For example, `abc\.\xyz` is a perfectly legal file name on linux, so a file with such name synced well so far and would have crashed the sync after this change is merged. I don't think crashing it would be good in such case.

To fix it, I think we need to a separate regular expression for Windows and non-windows systems.

The absolute matcher is also affected.

Reply:
Separate regular expressions will lead to further security vulnerabilities.  Using separate OS specific filters assumes the b2 file names only contain the local OS path separators.  Path separator conversion is performed in other places throughout the code (which could change any time in the future).  By dividing the regular expressions into platform-specific lists, malicious paths in a different platform's format will bypass the local-specific check and then be converted into a locally-malicious path later in the code.  Keeping the expressions together covers all the cases in one place.

Concerning the use of the "." path, the design is to disallow relative paths.  "." is a relative path and is not needed, nor desired, and would lead to sync problems.  Consider this list of perfectly legal files in b2:

"a/b/file.txt"
"a/b/./file.txt"

This example would not sync properly and should be disallowed.

Reply:
I just checked and the old sync could not have synced those names due to the conversions you've mentioned. 👍